expression/builtin_encryption.go

```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
 	return d, nil
 }
 
-type uncompressFunctionClass struct {
-	baseFunctionClass
-}
-
-func (c *uncompressFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinUncompressSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
-}
-
-type builtinUncompressSig struct {
-	baseBuiltinFunc
+// deflate compresses a string using the DEFLATE format.
+func deflate(data []byte) ([]byte, error) {
+	var buffer bytes.Buffer
+	w := zlib.NewWriter(&buffer)
+	if _, err := w.Write([]byte(data)); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := w.Close(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return buffer.Bytes(), nil
 }
 
-// uncompress is used for uncompress a string compiled with a compression library such as zlib.
-func uncompress(compressStr []byte) ([]byte, error) {
+// inflate uncompresses a string using the DEFLATE format.
+func inflate(compressStr []byte) ([]byte, error) {
 	reader := bytes.NewReader(compressStr)
 	var out bytes.Buffer
 	r, err := zlib.NewReader(reader)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	io.Copy(&out, r)
+	_, err = io.Copy(&out, r)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
 	r.Close()
 	return out.Bytes(), nil
 }
 
-// eval evals a builtinUncompressSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompress
-func (b *builtinUncompressSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+type compressFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *compressFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString)
 	if err != nil {
-		return d, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	arg := args[0]
-	if arg.IsNull() {
-		return d, nil
+	srcLen := args[0].GetType().Flen
+	compressBound := srcLen + (srcLen >> 12) + (srcLen >> 14) + (srcLen >> 25) + 13
```

Comment:
MySQL's `flen` seems to be `12+len*3.6`. However I didn't find magic number `3.6` in source code and I'm not sure about this.

I ported compressBound from zlib's `compressBound` function, which gives an upper bound of the length of compressed data: https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/compress.c#L81